models/db/common.go

```diff
@@ -20,3 +28,332 @@ func BuildCaseInsensitiveLike(key, value string) builder.Cond {
+	if len(uniqueCols) == 0 {
+		return false, fmt.Errorf("provided bean has no unique constraints")
+	}
+
+	var insertArgs []any
+
+	switch {
+	case setting.Database.UseSQLite3 || setting.Database.UsePostgreSQL || setting.Data
+		insertArgs = generateInsertNoConflictSQLAndArgs(tableName, colNames, args, autoIncrCol)
+	case setting.Database.UseMSSQL:
+		insertArgs = generateInsertNoConflictSQLAndArgsForMSSQL(tableName, colNames, args, uniqueCols, uniqueArgs, autoIncrCol)
+	default:
+		return false, fmt.Errorf("database type not supported")
+	}
+
+	if autoIncrCol != nil && (setting.Database.UsePostgreSQL || setting.Database.UseMSSQL) {
+		// Postgres and MSSQL do not use the LastInsertID mechanism
+		// Therefore use query rather than exec and read the last provided ID back in
+
+		res, err := e.Query(insertArgs...)
+		if err != nil {
+			return false, fmt.Errorf("error in query: %s, %w", insertArgs[0], err)
+		}
+		if len(res) == 0 {
+			// this implies there was a conflict
+			return false, nil
+		}
+
+		aiValue, err := table.AutoIncrColumn().ValueOf(bean)
+		if err != nil {
+			log.Error("unable to get value for autoincrcol of %#v %v", bean, err)
+		}
+
+		if aiValue == nil || !aiValue.IsValid() || !aiValue.CanSet() {
+			return true, nil
+		}
+
+		id := res[0][autoIncrCol.Name]
+		err = convert.AssignValue(*aiValue, id)
+		if err != nil {
+			return true, fmt.Errorf("error in assignvalue %v %v %w", id, res, err)
+		}
+		return true, nil
+	}
+
+	res, err := e.Exec(args...)
+	if err != nil {
+		return false, err
+	}
+
+	n, err := res.RowsAffected()
+	if err != nil {
+		return n != 0, err
+	}
+
+	if n != 0 && autoIncrCol != nil {
+		id, err := res.LastInsertId()
+		if err != nil {
+			return true, err
+		}
+		reflect.ValueOf(bean).Elem().FieldByName(autoIncrCol.FieldName).SetInt(id)
+	}
+
+	return n != 0, err
+}
+
+// generateInsertNoConflictSQLAndArgs will create the correct insert code for most of the DBs except MSSQL
+func generateInsertNoConflictSQLAndArgs(tableName string, colNames []string, args []any, autoIncrCol *schemas.Column) (insertArgs []any) {
+	sb := &strings.Builder{}
+
+	quote := x.Dialect().Quoter().Quote
+	write := func(args ...string) {
+		for _, arg := range args {
+			_, _ = sb.WriteString(arg)
+		}
+	}
+	write("INSERT ")
+	if setting.Database.UseMySQL && autoIncrCol == nil {
+		write("IGNORE ")
+	}
+	write("INTO ", quote(tableName), " (")
+	_ = x.Dialect().Quoter().JoinWrite(sb, colNames, ",")
+	write(") VALUES (?")
+	for range colNames[1:] {
+		write(",?")
+	}
+	switch {
+	case setting.Database.UsePostgreSQL:
+		write(") ON CONFLICT DO NOTHING")
+		if autoIncrCol != nil {
+			write(" RETURNING ", quote(autoIncrCol.Name))
+		}
+	case setting.Database.UseSQLite3:
+		write(") ON CONFLICT DO NOTHING")
+	case setting.Database.UseMySQL:
+		if autoIncrCol != nil {
+			write(") ON DUPLICATE KEY UPDATE ", quote(autoIncrCol.Name), " = ", quote(autoIncrCol.Name))
+		}
```

Comment:
Sorry to bother, but why `INSERT IGNORE` doesn't work for `autoIncrCol != nil`?

If you meant to do it for `LastInsertId`, I think it doesn't work.

Otherwise I really can not understand its purpose.

Reply:
Why don't you try it. When I wrote this code it required the ON DUPLICATE KEY UPDATE and not the IGNORE

Reply:
BECAUSE I HAVE TRIED


![image](https://user-images.githubusercontent.com/2114189/217302383-264e0f4c-123c-4708-9f0c-db151bdac98b.png)

Reply:
Table:

```sql
CREATE TABLE `t` (
  `id` int NOT NULL AUTO_INCREMENT,
  `k` varchar(100),
  `v` varchar(100) DEFAULT '',
  `i` int DEFAULT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `k` (`k`)
) ENGINE=InnoDB AUTO_INCREMENT=9
```

Reply:
What you really need is this:

`update id=last_insert_id(id)`


```
mysql> insert into t (k) values ('1') on duplicate key update id=last_insert_id(id);
Query OK, 0 rows affected (0.00 sec)

mysql> select last_insert_id();
+------------------+
| last_insert_id() |
+------------------+
|                1 |
+------------------+
1 row in set (0.00 sec)

```

Reply:
a) I would have thought that last_insert_id() requires that there are no other id insertions in the meantime.
b) id = id works
c) IGNORE is only used when there is no autoincrement ID present in the table where there is no id to update.
d) I've written some tests now.

Reply:
```
mysql> insert into t (k) values ('a');
Query OK, 1 row affected (0.00 sec)

mysql> select last_insert_id();
+------------------+
| last_insert_id() |
+------------------+
|                1 |
+------------------+
1 row in set (0.00 sec)

mysql> insert into t (k) values ('b');
Query OK, 1 row affected (0.00 sec)

mysql> select last_insert_id();
+------------------+
| last_insert_id() |
+------------------+
|                2 |
+------------------+
1 row in set (0.00 sec)

mysql> insert into t (k) values ('a') on duplicate key update id=id;
Query OK, 0 rows affected (0.00 sec)

!!!!!!! HERE You are inserting duplicated `a` and want to see `id=1`, BUT:

mysql> select last_insert_id();
+------------------+
| last_insert_id() |
+------------------+
|                2 |
+------------------+
1 row in set (0.00 sec)

mysql> select * from t where id=2;
+----+------+------+------+
| id | k    | v    | i    |
+----+------+------+------+
|  2 | b    |      | NULL |
+----+------+------+------+
1 row in set (0.00 sec)

```